### PR TITLE
PaywallViewController: option to not automatically dismiss and add `onRequestedDismissal` 

### PR DESCRIPTION
**Pull request by vegaro**
Please read https://github.com/RevenueCat/react-native-purchases/pull/913

### CODE REVIEW

File: RevenueCatUI/PaywallView.swift
Comment (tonidero) [inline RevenueCatUI/PaywallView.swift:281]: Hmm maybe I'm wrong but I think we could make this not nullable now? (We always pass a value I think)

**Review by tonidero — APPROVED** [review]
LGTM!

File: RevenueCatUI/PaywallView.swift
Comment (vegaro) [reply RevenueCatUI/PaywallView.swift:281]: there will only be a value if `onRequestedDismissal` on PaywallView is called, and it's not even a public function (but PaywallView is), so this needs to be nullable

Comment (tonidero) [reply RevenueCatUI/PaywallView.swift:281]: Ohh ok makes sense

File: RevenueCatUI/UIKit/PaywallViewController.swift
Comment (joshdholtz) [reply RevenueCatUI/UIKit/PaywallViewController.swift:303]: I think I agree with this too ☝️ We should only set this if the delegate is set? Or... manually dismiss this if the delegate isn't set.

Comment (vegaro) [reply RevenueCatUI/UIKit/PaywallViewController.swift]: yes, you both are right. Made some changes in https://github.com/RevenueCat/purchases-ios/pull/3738/commits/9efc7252f4b231b4a67b5296c3dbc273ee497588. Let me know if they make sense. It looks like I broke something though, but looking into it.

File: RevenueCatUI/UIKit/PaywallViewController.swift
Comment (vegaro) [reply RevenueCatUI/UIKit/PaywallViewController.swift]: done!

File: RevenueCatUI/UIKit/PaywallFooterViewController.swift
Comment (tonidero) [inline RevenueCatUI/UIKit/PaywallFooterViewController.swift:38]: Hmm I think we should still accept a dismissalRequest parameter for the footer, since there might be a successful purchase? Alternatively, we can just devs use the purchaseCompleted callback, which would be called in the same location... But not sure if we want to be consistent with that behavior between the full screen and footer.

File: RevenueCatUI/UIKit/PaywallFooterViewController.swift
Comment (tonidero) [inline RevenueCatUI/UIKit/PaywallFooterViewController.swift:65]: Hmm while I think we need to add it to the other constructors here, why do we add it to the unavailable constructor?

**Review by tonidero — COMMENTED** [review]
Just that question about the footer. Otherwise looks good!

File: RevenueCatUI/UIKit/PaywallFooterViewController.swift
Comment (vegaro) [reply RevenueCatUI/UIKit/PaywallFooterViewController.swift]: oh actually, this one is overriden from `PaywallViewController`, so it's required :/

File: RevenueCatUI/UIKit/PaywallFooterViewController.swift
Comment (vegaro) [reply RevenueCatUI/UIKit/PaywallFooterViewController.swift]: hmmm, yes. I think it makes sense, I will add it to `PaywallFooterViewController`

Comment (vegaro) [reply RevenueCatUI/UIKit/PaywallFooterViewController.swift]: Ok... so I added it, but the callback is not being called because there's this code in `PaywallView`

```
              if self.mode.isFullScreen, purchased {
                    Logger.debug(Strings.dismissing_paywall)
                    guard let onRequestedDismissal = self.onRequestedDismissal else {
                        self.dismiss()
                        return
                    }
                    onRequestedDismissal()
                }
```

I am going to see what happens in Android... but if we don't call the `onDismiss` in Android, I think we shouldn't do it here either

**Comment by vegaro** [comment]
@joshdholtz can you take a look? I can't merge until I get your approval 😄

**Review by joshdholtz — APPROVED** [review]
LGTM 💪

**Pull request merged**
